fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/InstrumentedReadTransaction.java

```diff
@@ -0,0 +1,249 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.KeySelector;
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.Range;
+import com.apple.foundationdb.ReadTransaction;
+import com.apple.foundationdb.StreamingMode;
+import com.apple.foundationdb.TransactionOptions;
+import com.apple.foundationdb.async.AsyncIterable;
+import com.apple.foundationdb.async.AsyncIterator;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.provider.common.StoreTimer;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+
+abstract class InstrumentedReadTransaction<T extends ReadTransaction> implements ReadTransaction {
+    @Nonnull
+    protected StoreTimer timer;
+    @Nonnull
+    protected T underlying;
+
+    public InstrumentedReadTransaction(@Nonnull StoreTimer timer, @Nonnull T underlying) {
+        this.timer = timer;
+        this.underlying = underlying;
+    }
+
+    @Override
+    public boolean isSnapshot() {
+        return underlying.isSnapshot();
+    }
+
+    @Override
+    public CompletableFuture<Long> getReadVersion() {
+        return underlying.getReadVersion();
+    }
+
+    @Override
+    public void setReadVersion(long l) {
+        underlying.setReadVersion(l);
+    }
+
+    @Override
+    public boolean addReadConflictRangeIfNotSnapshot(byte[] keyBegin, byte[] keyEnd) {
+        return underlying.addReadConflictRangeIfNotSnapshot(keyBegin, keyEnd);
+    }
+
+    @Override
+    public boolean addReadConflictKeyIfNotSnapshot(byte[] key) {
+        return underlying.addReadConflictKeyIfNotSnapshot(key);
+    }
+
+    @Override
+    public CompletableFuture<byte[]> get(byte[] key) {
+        timer.increment(FDBStoreTimer.Counts.READS);
+        return underlying.get(key).thenApply(this::recordRead);
+    }
+
+    @Override
+    public CompletableFuture<byte[]> getKey(KeySelector keySelector) {
+        timer.increment(FDBStoreTimer.Counts.READS);
+        return underlying.getKey(keySelector).thenApply(this::recordRead);
+    }
+
+    @Override
+    public AsyncIterable<KeyValue> getRange(KeySelector begin, KeySelector end) {
+        /* Should this could as one read? */
```

Comment:
Good question. I think (?) that the FDB metric that goes into machine readable status is 1 range read = 1 read, so if we wanted to be able to correlated, say, reads from this timer to reads from that timer, we'd want them to be the same. This also means that we count work from empty range reads, which we may care about. I suppose we still know how many entries get back from our other, higher level metrics (like `FDBStoreTimer.Counts.LOAD_RECORD`). Hmm.